File: bench/dimer_P_tau/plot_P.py

import glob
import numpy as np
import pandas as ps

import h5py

import matplotlib.pyplot as plt

from triqs_tprf.ParameterCollection import ParameterCollection

filenames = [
'data_dimer_ntau_8192_maxorder_1_Nsamples_256_md5_3ec769d56f60d2d77b6883842c5ef185.h5',
'data_dimer_ntau_32768_maxorder_3_Nsamples_4096_md5_72c2152b0d31e6f8751c034569e10ba7.h5',
'data_dimer_ntau_32768_maxorder_4_Nsamples_4096_md5_991d9643db7885b6e65f46e3b1f98b7c.h5'
    ]

ds = []
for filename in filenames:
    print(f'--> Loading: {filename}')
    d = {}
    with h5py.File(filename, 'r') as a:
        for key, value in a['data'].attrs.items():
            d[key] = [value]
        for key, value in a['data'].items():
            d[key] = np.array(value)
    d = ParameterCollection(**d)
    ds.append(d)

plt.figure()
subp = [3, 1, 1]

plt.subplot(*subp); subp[-1] += 1
for d in ds:
    plt.plot(d.tau, d.P1, 'r', label=f'P1 order {d.maxorder}')
    plt.plot(d.tau, d.P2, 'g', label=f'P2 order {d.maxorder}')
    plt.plot(d.tau, d.P1_exact, 'c--')
    plt.plot(d.tau, d.P2_exact, 'm--')

plt.subplot(*subp); subp[-1] += 1
for d in ds:
    plt.plot(d.tau, d.P1 - d.P1_exact, label=f'P1 order {d.maxorder}')
    plt.plot(d.tau, d.P2 - d.P2_exact, label=f'P2 order {d.maxorder}')

plt.subplot(*subp); subp[-1] += 1
for d in ds:
    plt.plot(d.tau, np.abs(d.P1 - d.P1_exact), label=f'P1 order {d.maxorder}')
    plt.plot(d.tau, np.abs(d.P2 - d.P2_exact), label=f'P2 order {d.maxorder}')
plt.semilogy([], [])

plt.legend(loc='best')
plt.tight_layout()
plt.show()

File: bench/dimer_P_tau/plot_P.jl
# QInchworm.jl
#
# Copyright (C) 2021-2023 I. Krivenko, H. U. R. Strand and J. Kleinhenz
#
# QInchworm.jl is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# QInchworm.jl is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# QInchworm.jl. If not, see <http://www.gnu.org/licenses/>.
#
# Authors: Hugo U. R. Strand, Igor Krivenko

using HDF5; h5 = HDF5

using PyPlot; plt = PyPlot

filenames = filter(f -> endswith(f, ".h5"), readdir(".", join=true))

ds = []
for filename in filenames
    println("--> Loading: $filename")
    d = Dict()
    h5.h5open(filename, "r") do fid
        grp = fid["data"]
        for key in keys(h5.attributes(grp))
            d[key] = h5.read_attribute(grp, key)
        end
        for key in keys(grp)
            d[key] = h5.read(grp, key)
        end
    end
    push!(ds, d)
end

plt.figure()
subp = [3, 1, 1]

plt.subplot(subp...); subp[end] += 1
for d in ds
    maxorder = d["maxorder"][1]
    plt.plot(d["tau"], d["P1"], "r", label="P1 order $maxorder")
    plt.plot(d["tau"], d["P2"], "g", label="P2 order $maxorder")
    plt.plot(d["tau"], d["P1_exact"], "c--")
    plt.plot(d["tau"], d["P2_exact"], "m--")
end

plt.subplot(subp...); subp[end] += 1
for d in ds
    maxorder = d["maxorder"][1]
    plt.plot(d["tau"], d["P1"] - d["P1_exact"], label="P1 order $maxorder")
    plt.plot(d["tau"], d["P2"] - d["P2_exact"], label="P2 order $maxorder")
end

plt.subplot(subp...); subp[end] += 1
for d in ds
    maxorder = d["maxorder"][1]
    plt.plot(d["tau"], abs.(d["P1"] - d["P1_exact"]), label="P1 order $maxorder")
    plt.plot(d["tau"], abs.(d["P2"] - d["P2_exact"]), label="P2 order $maxorder")
end

plt.semilogy([], [])
plt.legend(loc="best")
plt.tight_layout()
plt.show()
